content/develop/concepts/architecture/architecture.md

```diff
@@ -23,3 +23,24 @@ Keep in mind the following considerations when building your Streamlit app:
 - Your app will not have access to a user's files, directories, or OS. Your app can only work with specific files a user has uploaded to your app through a widget like `st.file_uploader`.
 - If your app communicates with any peripheral devices (like cameras), you must use Streamlit commands or custom components that will access those devices _through the user's browser_ and correctly communicate between the client (frontend) and server (backend).
 - If your app opens or uses any program or process outside of Python, they will run on the server. For example, you may want to use `webrowser` to open a browser for the user, but this will not work as expected when viewing your app over a network; it will open a browser on the Streamlit server, unseen by the user.
+- If you use load balancing or replication in your deployment, some Streamlit features won't work without session affinity or stickiness. For more information, continue reading.
+
+## WebSockets and session management
+
+While most Streamlit app developers don’t need to interact directly with WebSockets, understanding their role is important for advanced deployments, custom components, or managing connections at scale.
+
+Streamlit’s server is built on the Tornado web framework, which uses WebSockets to maintain a persistent, two-way communication channel between the client and server. This persistent connection allows the server to push real-time updates to the client and maintain session context for each user. Each browser tab or window creates its own WebSocket connection, resulting in a separate session within your app.
+
+In large-scale or production deployments, load balancing and server replication are common. However, WebSockets require special consideration in these environments. When a client requests media (such as an image or audio file) via HTTP, there is no session context attached to that request. In deployments with multiple server replicas or load balancers, the HTTP request for the media file might be routed to a different server than the one handling the user’s WebSocket connection. If the media file isn’t available on all replicas, you may encounter errors like `MediaFileStorageError: Bad filename`. Any command that allows the user to upload files can also be impacted and may raise HTTP status code 400.
+
+### Session affinity or stickiness
+
+In general, you can do one of the following to resolve or reduce this limitation:
+
+- Enable session affinity (also known as stickiness).
+- Convert media to a Base64 encoded data URI before passing it to a Streamlit command.
+- Save dynamically generated files to a stable location outside of your server replicas, and pass references to that stable location.
```

Comment:
Maybe get a bit more specific:
 Save dynamically generated files to a stable location outside of your server replicas (e.g. S3 storage), and pass the URL to the file on the external storage.